File: starters/react-native/src/api/typed-api-client.ts
import {
  createMobileRequestContext,
  mapCanonicalMobileError,
  type CanonicalMobileError,
  type MobileRequestContext,
} from '../platform/runtime-contract';

export interface MobileTransportResponse<T> {
  readonly ok: boolean;
  readonly status: number;
  json(): Promise<T>;
}

export type MobileTransport = (
  path: string,
  init: Readonly<{ method: string; headers: Readonly<Record<string, string>> }>,
) => Promise<MobileTransportResponse<unknown>>;

export class MobileApiError extends Error {
  constructor(readonly response: CanonicalMobileError) {
    super(response.message);
    this.name = 'MobileApiError';
  }
}

export class TypedMobileApiClient {
  constructor(
    private readonly transport: MobileTransport,
    private readonly contextFactory: () => MobileRequestContext = () =>
      createMobileRequestContext({
        requestId: 'mobile-request',
        nextTraceId: () => '1'.repeat(32),
        nextSpanId: () => '2'.repeat(16),
      }),
  ) {}

  async request<T>(path: string, method = 'GET'): Promise<T> {
    const context = this.contextFactory();
    const response = await this.transport(path, {
      method,
      headers: {
        'X-Request-Id': context.requestId,
        traceparent: context.traceparent,
      },
    });
    const body = await response.json();
    if (!response.ok) throw new MobileApiError(mapCanonicalMobileError(body));
    return body as T;
  }
}

File: starters/flutter/lib/src/core/api/typed_api_client.dart
import '../platform/runtime_contract.dart';

final class MobileTransportResponse {
  const MobileTransportResponse({
    required this.ok,
    required this.statusCode,
    required this.body,
  });

  final bool ok;
  final int statusCode;
  final Object? body;
}

typedef MobileTransport =
    Future<MobileTransportResponse> Function(
      String path,
      String method,
      Map<String, String> headers,
    );

final class TypedMobileApiClient {
  const TypedMobileApiClient({
    required MobileTransport transport,
    required MobileRequestContext Function() contextFactory,
  }) : _transport = transport,
       _contextFactory = contextFactory;

  final MobileTransport _transport;
  final MobileRequestContext Function() _contextFactory;

  Future<T> request<T>(
    String path, {
    String method = 'GET',
    required T Function(Object? body) decode,
  }) async {
    final context = _contextFactory();
    final response = await _transport(path, method, {
      'X-Request-Id': context.requestId,
      'traceparent': context.traceparent,
    });
    if (!response.ok) throw CanonicalMobileError.fromJson(response.body);
    return decode(response.body);
  }
}
